File: server/src/routes/profiles.ts
import { Router } from "express";
import { z } from "zod";

import { profileRepo } from "../db.js";
import { browserManager } from "../services/browser.js";

const router = Router();

const profileCreateSchema = z.object({
  name: z.string().min(1).max(120),
  site_domain: z.string().min(1).max(255),
  description: z.string().nullable().optional(),
});

const profileUpdateSchema = z.object({
  name: z.string().min(1).max(120).optional(),
  site_domain: z.string().min(1).max(255).optional(),
  description: z.string().nullable().optional(),
});

const loginStartSchema = z.object({
  start_url: z.string().url().optional(),
  use_chrome_cdp: z.boolean().optional(),
});

const profileResolveSchema = z.object({
  url: z.string().url(),
});

function storageExists(profileId: number): boolean {
  return browserManager.profileHasStorage(profileId);
}

function parseSiteDomain(url: string): string {
  return new URL(url).hostname.toLowerCase();
}

router.get("", (_req, res) => {
  res.json(profileRepo.list());
});

router.post("/resolve", (req, res) => {
  const parsed = profileResolveSchema.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ detail: parsed.error.flatten() });
    return;
  }

  const siteDomain = parseSiteDomain(parsed.data.url);
  const existing = profileRepo.getBySiteDomain(siteDomain);
  if (existing) {
    res.json({
      profile: existing,
      created: false,
      has_storage: storageExists(existing.id),
      site_domain: siteDomain,
    });
    return;
  }

  const profile = profileRepo.create({
    name: siteDomain,
    site_domain: siteDomain,
    description: `自动创建于一键配置（${parsed.data.url}）`,
  });

  res.status(201).json({
    profile,
    created: true,
    has_storage: false,
    site_domain: siteDomain,
  });
});

router.post("", (req, res) => {
  const parsed = profileCreateSchema.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ detail: parsed.error.flatten() });
    return;
  }

  if (profileRepo.getByName(parsed.data.name)) {
    res.status(400).json({ detail: "配置档名称已存在" });
    return;
  }

  const profile = profileRepo.create(parsed.data);
  res.status(201).json(profile);
});

router.get("/:profileId", (req, res) => {
  const profileId = Number(req.params.profileId);
  const profile = profileRepo.get(profileId);
  if (!profile) {
    res.status(404).json({ detail: "配置档不存在" });
    return;
  }
  res.json(profile);
});

router.patch("/:profileId", (req, res) => {
  const profileId = Number(req.params.profileId);
  const parsed = profileUpdateSchema.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ detail: parsed.error.flatten() });
    return;
  }

  const profile = profileRepo.update(profileId, parsed.data);
  if (!profile) {
    res.status(404).json({ detail: "配置档不存在" });
    return;
  }
  res.json(profile);
});

router.delete("/:profileId", (req, res) => {
  const profileId = Number(req.params.profileId);
  if (!profileRepo.delete(profileId)) {
    res.status(404).json({ detail: "配置档不存在" });
    return;
  }
  res.status(204).send();
});

router.post("/:profileId/login/start", async (req, res) => {
  const profileId = Number(req.params.profileId);
  const profile = profileRepo.get(profileId);
  if (!profile) {
    res.status(404).json({ detail: "配置档不存在" });
    return;
  }

  const parsed = loginStartSchema.safeParse(req.body ?? {});
  if (!parsed.success) {
    res.status(400).json({ detail: parsed.error.flatten() });
    return;
  }

  const startUrl = parsed.data.start_url ?? `https://${profile.site_domain}`;

  browserManager.reconcileLoginSession(profileId);

  try {
    await browserManager.startLoginSession(
      profileId,
      startUrl,
      parsed.data.use_chrome_cdp ?? false,
    );
  } catch (error) {
    res.status(409).json({ detail: error instanceof Error ? error.message : String(error) });
    return;
  }

  profileRepo.updateLoginStatus(profileId, "logging_in");
  res.json({
    profile_id: profileId,
    status: "active",
    message: parsed.data.use_chrome_cdp
      ? `已在你的 Chrome 中打开 ${startUrl}，登录后点击「保存登录状态」`
      : `已打开浏览器窗口，请手动登录 ${profile.site_domain}，完成后点击「保存登录状态」`,
  });
});

router.post("/:profileId/import-chrome", async (req, res) => {
  const profileId = Number(req.params.profileId);
  const profile = profileRepo.get(profileId);
  if (!profile) {
    res.status(404).json({ detail: "配置档不存在" });
    return;
  }

  try {
    const storagePath = await browserManager.importChromeLogin(profileId);
    const updated = profileRepo.updateLoginStatus(profileId, "logged_in", storagePath);
    res.json({
      profile: updated,
      message: `已从本机 Chrome 导入 ${profile.site_domain} 的登录环境`,
    });
  } catch (error) {
    res.status(400).json({ detail: error instanceof Error ? error.message : String(error) });
  }
});

router.post("/:profileId/login/save", async (req, res) => {
  const profileId = Number(req.params.profileId);
  const profile = profileRepo.get(profileId);
  if (!profile) {
    res.status(404).json({ detail: "配置档不存在" });
    return;
  }

  try {
    const storagePath = await browserManager.saveLoginSession(profileId);
    const updated = profileRepo.updateLoginStatus(profileId, "logged_in", storagePath);
    res.json(updated);
  } catch (error) {
    res.status(400).json({ detail: error instanceof Error ? error.message : String(error) });
  }
});

router.post("/:profileId/login/cancel", async (req, res) => {
  const profileId = Number(req.params.profileId);
  const profile = profileRepo.get(profileId);
  if (!profile) {
    res.status(404).json({ detail: "配置档不存在" });
    return;
  }

  await browserManager.cancelLoginSession(profileId);
  profileRepo.updateLoginStatus(
    profileId,
    storageExists(profileId) ? "logged_in" : "unknown",
  );

  res.json({
    profile_id: profileId,
    status: "cancelled",
    message: "已取消登录会话",
  });
});

export default router;

File: app/api/profiles.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import settings
from app.database import get_session
from app.models import Profile
from app.schemas import (
    LoginSessionRead,
    LoginStartRequest,
    ProfileCreate,
    ProfileRead,
    ProfileUpdate,
)
from app.services.browser import browser_manager

router = APIRouter(prefix="/api/profiles", tags=["profiles"])


def _storage_exists(profile_id: int) -> bool:
    return browser_manager.profile_storage_path(profile_id).exists()


@router.get("", response_model=list[ProfileRead])
async def list_profiles(session: AsyncSession = Depends(get_session)) -> list[Profile]:
    result = await session.execute(select(Profile).order_by(Profile.id.desc()))
    return list(result.scalars().all())


@router.post("", response_model=ProfileRead, status_code=status.HTTP_201_CREATED)
async def create_profile(
    payload: ProfileCreate,
    session: AsyncSession = Depends(get_session),
) -> Profile:
    existing = await session.execute(select(Profile).where(Profile.name == payload.name))
    if existing.scalar_one_or_none():
        raise HTTPException(status_code=400, detail="配置档名称已存在")

    profile = Profile(**payload.model_dump(), login_status="unknown")
    session.add(profile)
    await session.commit()
    await session.refresh(profile)
    return profile


@router.get("/{profile_id}", response_model=ProfileRead)
async def get_profile(
    profile_id: int,
    session: AsyncSession = Depends(get_session),
) -> Profile:
    profile = await session.get(Profile, profile_id)
    if profile is None:
        raise HTTPException(status_code=404, detail="配置档不存在")
    return profile


@router.patch("/{profile_id}", response_model=ProfileRead)
async def update_profile(
    profile_id: int,
    payload: ProfileUpdate,
    session: AsyncSession = Depends(get_session),
) -> Profile:
    profile = await session.get(Profile, profile_id)
    if profile is None:
        raise HTTPException(status_code=404, detail="配置档不存在")

    for key, value in payload.model_dump(exclude_unset=True).items():
        setattr(profile, key, value)

    await session.commit()
    await session.refresh(profile)
    return profile


@router.delete("/{profile_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_profile(
    profile_id: int,
    session: AsyncSession = Depends(get_session),
) -> None:
    profile = await session.get(Profile, profile_id)
    if profile is None:
        raise HTTPException(status_code=404, detail="配置档不存在")

    profile_dir = settings.profiles_dir / str(profile_id)
    await session.delete(profile)
    await session.commit()

    if profile_dir.exists():
        for item in profile_dir.glob("*"):
            item.unlink(missing_ok=True)
        profile_dir.rmdir()


@router.post("/{profile_id}/login/start", response_model=LoginSessionRead)
async def start_login(
    profile_id: int,
    payload: LoginStartRequest,
    session: AsyncSession = Depends(get_session),
) -> LoginSessionRead:
    profile = await session.get(Profile, profile_id)
    if profile is None:
        raise HTTPException(status_code=404, detail="配置档不存在")

    start_url = str(payload.start_url) if payload.start_url else f"https://{profile.site_domain}"

    try:
        await browser_manager.start_login_session(profile_id, start_url)
    except RuntimeError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc

    profile.login_status = "logging_in"
    await session.commit()

    return LoginSessionRead(
        profile_id=profile_id,
        status="active",
        message=f"已打开浏览器窗口，请手动登录 {profile.site_domain}，完成后点击「保存登录状态」",
    )


@router.post("/{profile_id}/login/save", response_model=ProfileRead)
async def save_login(
    profile_id: int,
    session: AsyncSession = Depends(get_session),
) -> Profile:
    profile = await session.get(Profile, profile_id)
    if profile is None:
        raise HTTPException(status_code=404, detail="配置档不存在")

    try:
        storage_path = await browser_manager.save_login_session(profile_id)
    except RuntimeError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    profile.storage_state_path = str(storage_path)
    profile.login_status = "logged_in"
    await session.commit()
    await session.refresh(profile)
    return profile


@router.post("/{profile_id}/login/cancel", response_model=LoginSessionRead)
async def cancel_login(
    profile_id: int,
    session: AsyncSession = Depends(get_session),
) -> LoginSessionRead:
    profile = await session.get(Profile, profile_id)
    if profile is None:
        raise HTTPException(status_code=404, detail="配置档不存在")

    await browser_manager.cancel_login_session(profile_id)
    profile.login_status = "logged_in" if _storage_exists(profile_id) else "unknown"
    await session.commit()

    return LoginSessionRead(
        profile_id=profile_id,
        status="cancelled",
        message="已取消登录会话",
    )
